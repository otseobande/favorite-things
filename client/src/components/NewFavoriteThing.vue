<template>
  <div class="text-gray-700">
    <h2 class="text-lg font-bold">
        <span v-if='editting'>
          <span>Edit </span>
          <span v-if="newFavoriteThing.title">{{newFavoriteThing.title}}</span>
          <span v-else>favorite thing</span>
        </span>
        <span v-else>
          Add new favorite <span class="lowercase">{{singularizedCategoryName}}</span>
        </span>
    </h2>
    <form class="mt-5" @submit.prevent="handleAddNewFavoriteThing">
      <div>
        <label class="font-bold">Title</label>
        <t-input
          v-model="$v.newFavoriteThing.title.$model"
          name="title"
          class="h-10 w-full"
          :status="!$v.newFavoriteThing.title.$dirty
            ? null
            : $v.newFavoriteThing.title.required && $v.newFavoriteThing.title.maxLength"
          :placeholder="`Enter favorite ${singularizedCategoryName} title`"
        />
        <div v-if="$v.newFavoriteThing.title.$dirty">
          <div
            v-if="!$v.newFavoriteThing.title.required"
            class="text-sm text-red-500"
          >
            Title is required
          </div>
          <div
            v-if="!$v.newFavoriteThing.title.maxLength"
            class="text-sm text-red-500"
          >
            Title should be less than 30 characters.
          </div>
        </div>
      </div>
      <div class="mt-2">
        <label class="font-bold">Description </label>
        <span class="text-sm italic text-gray-500">(optional)</span>
        <t-textarea
          v-model="$v.newFavoriteThing.description.$model"
          class="w-full"
          name="description"
          :status="!$v.newFavoriteThing.description.$dirty ?
            null :
            $v.newFavoriteThing.description.minLength"
          :placeholder="`Describe favorite ${singularizedCategoryName}`"
        />
        <div
          v-if="$v.newFavoriteThing.description.$dirty &&
            $v.newFavoriteThing.description.minLength === false"
          class="text-sm text-red-500"
        >
          Description should be greater than 10 characters.
        </div>
      </div>
      <div class="mt-2">
        <label class="font-bold">Ranking</label>
        <div>
          <t-input
            v-model="$v.newFavoriteThing.ranking.$model"
            class="h-10 mr-0 inline"
            name="ranking"
            type="number"
            min="1"
            :max="`${this.category.favorite_things.length + 1}`"
            :status="!$v.newFavoriteThing.ranking.$dirty
              ? null
              : (
                $v.newFavoriteThing.ranking.required &&
                $v.newFavoriteThing.ranking.betweenRanking
              )"
          />
          <small class="ml-2 capitalize">{{rankingInfo}}</small>
        </div>
        <div v-if="$v.newFavoriteThing.ranking.$dirty">
          <div
            v-if="!$v.newFavoriteThing.ranking.required"
            class="text-sm text-red-500"
          >
            Ranking is required
          </div>
          <div
            v-if="!$v.newFavoriteThing.ranking.betweenRanking"
            class="text-sm text-red-500"
          >
            Ranking is out of range
          </div>
        </div>
      </div>
      <AdditionalInformation
        @dataChange="handleMetadataChange"
        :currentMetadata="transformMetadata()"
        editting
      />
      <div class="mt-8 flex">
        <t-button
          :disabled="submitting"
          variant="primary"
          size="sm"
          class="mr-2 font-bold"
          type='submit'
          v-if="editting"
          @click.prevent="handleEditNewFavoriteThing"
        >
          <span v-if="submitting">
            <img src="../assets/spinner.svg" width="25">
          </span>
          <span v-else>
            Update<span class="lowercase">&nbsp;{{singularizedCategoryName}}</span>
          </span>
        </t-button>
        <t-button
          :disabled="submitting"
          variant="primary"
          size="sm"
          class="mr-2 font-bold"
          type='submit'
          v-else
        >
          <span v-if="submitting">
            <img src="../assets/spinner.svg" width="25">
          </span>
          <span v-else>
              Save favorite<span class="lowercase">&nbsp;{{singularizedCategoryName}}</span>
          </span>
        </t-button>

        <t-button size="sm" class="font-bold" @click="emitFinished">
          Cancel
        </t-button>
      </div>
    </form>
  </div>
</template>

<script>
// Vuex
import { mapActions } from 'vuex';
import * as actionTypes from '../store/actions.type';

import { required, minLength, maxLength } from 'vuelidate/lib/validators';

// components
import AdditionalInformation from './AdditionalInformation.vue';

export default {
  props: {
    category: {
      type: Object,
      required: true
    },
    show: {
      type: Boolean,
      default: false
    },
    singularizedCategoryName: {
      type: String,
      required: true
    },
    favoriteThing: {
      type: Object,
    },
    editting: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      newFavoriteThing: this.getInitialNewFavoriteThing(),
      submitting: false,
    }
  },
  components: {
    AdditionalInformation
  },
  validations: {
    newFavoriteThing: {
      title: {
        required,
        maxLength: maxLength(30)
      },
      description: {
        minLength: minLength(4)
      },
      ranking: {
        required,
        betweenRanking: function(ranking) {
          return (
            ranking >= 1 &&
            ranking <= this.category.favorite_things.length + 1
          );
        }
      }
    }
  },
  methods: {
    ...mapActions([
      actionTypes.ADD_FAVORITE_THING,
      actionTypes.UPDATE_FAVORITE_THING,
    ]),
    getInitialNewFavoriteThing() {
      return {
        title: '',
        description: '',
        category: this.category.id,
        ranking: this.category.favorite_things.length + 1,
        metadata: []
      }
    },
    async handleAddNewFavoriteThing() {
      try{
        this.submitting = true;

        if(this.$v.newFavoriteThing.$invalid) {
          this.submitting = false;
          return this.$v.newFavoriteThing.$touch();
        }

        await this[actionTypes.ADD_FAVORITE_THING](this.newFavoriteThing);
        this.submitting = false;
        this.emitFinished();
      } finally {
        this.submitting = false;
      }
    },
    async handleEditNewFavoriteThing() {
      try{
        this.submitting = true;

        if(this.$v.newFavoriteThing.$invalid) {
          this.submitting = false;
          return this.$v.newFavoriteThing.$touch();
        }

        await this[actionTypes.UPDATE_FAVORITE_THING]({
          favoriteThingId: this.newFavoriteThing.id,
          updatedFavoriteThingData: this.newFavoriteThing
        });
        this.submitting = false;
        this.emitFinished();
      } finally {
        this.submitting = false;
      }
    },
    handleMetadataChange (value) {
      this.newFavoriteThing.metadata = value;
    },
    emitFinished() {
      this.$emit('finished');
      this.newFavoriteThing = this.getInitialNewFavoriteThing();
      this.$v.newFavoriteThing.$reset()
    },
    transformMetadata () {
      const metadata = typeof(this.newFavoriteThing.metadata) === 'object'
        ? this.newFavoriteThing.metadata
        : JSON.parse(this.newFavoriteThing.metadata);

      const transformedMetadata = metadata.map(data => {
        if (data.type === 'list' && Array.isArray(data.value)) {
          const updatedData = {
            ...data,
            value: [ ...data.value]
          }

          updatedData.value = updatedData.value.map(valueObject => {
            return {
              key: '',
              value: valueObject
            }
          });

          return updatedData;
        }

        return data;
      });

      return transformedMetadata;
    }
  },
  computed: {
    rankingInfo () {
      const currentRanking = Number(this.newFavoriteThing.ranking);
      
      if (currentRanking === 1) {
        return `Rank as favorite ${this.singularizedCategoryName}`
      }

      if (
        this.favoriteThing &&
        this.favoriteThing.ranking === currentRanking
      ) {
        return ''
      }

      if (
        currentRanking >= 1 &&
        currentRanking <= this.category.favorite_things.length
      ) {
          const newFavoriteThingInPostion = this.category.favorite_things[
            this.newFavoriteThing.ranking - 1
          ];

          return `Rank above ${newFavoriteThingInPostion.title}`
      }

      if (currentRanking === this.category.favorite_things.length + 1) {
        return `Rank as least favorite ${this.singularizedCategoryName}`
      }

      return '';
    },
  },

  mounted() {
    if (this.favoriteThing) {
      this.newFavoriteThing = this.favoriteThing;
    }
  }
}
</script>
