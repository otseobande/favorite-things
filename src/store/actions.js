import * as actionTypes from './actions.type';
import * as mutationTypes from './mutations.type';

import categoryService from '../services/categoryService';
import favoriteThingService from '../services/favoriteThingService';

export default {
  async [actionTypes.FETCH_CATEGORIES]({ commit }) {
    const response = await categoryService.get();

    commit(mutationTypes.SET_CATEGORIES, response.data.results)
  },

  async [actionTypes.ADD_FAVORITE_THING]({ commit }, newFavoriteThing) {
    const response = await favoriteThingService.create(newFavoriteThing);

    commit(mutationTypes.ADD_FAVORITE_THING, response.data)
  },

  async [actionTypes.UPDATE_FAVORITE_THING]({ commit }, {
    favoriteThingId,
    updatedFavoriteThingData
  }) {
    const response = await favoriteThingService.update({
      favoriteThingId,
      updatedFavoriteThingData
    })

    commit(mutationTypes.UPDATE_FAVORITE_THING, response.data);
  },

  async [actionTypes.DELETE_FAVORITE_THING]({ commit }, { favoriteThingId }) {
    await favoriteThingService.delete({
      favoriteThingId
    });

    commit(mutationTypes.DELETE_FAVORITE_THING, favoriteThingId)
  },

  async [actionTypes.ADD_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.ADD_CATEGORY, {
      id: 'template',
      name: '',
      favorite_things: []
    })
  },

  async [actionTypes.DELETE_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.DELETE_CATEGORY, 'template')
  },

  async [actionTypes.ADD_CATEGORY]({ commit }, newCategory) {
    const response = await categoryService.create(newCategory)

    commit(mutationTypes.ADD_CATEGORY, response.data)
  },

  async [actionTypes.DELETE_CATEGORY]({ commit }, { categoryId }) {
    await categoryService.delete(categoryId);

    commit(mutationTypes.DELETE_CATEGORY, categoryId);
  }
}
