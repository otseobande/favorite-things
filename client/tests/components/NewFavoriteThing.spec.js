import { shallowMount, createLocalVue } from '@vue/test-utils';
import NewFavoriteThing from '../../src/components/NewFavoriteThing.vue';
import VueTailwind from 'vue-tailwind';
import Vuex from 'vuex';
import actions from '../../src/store/actions';
import Vuelidate from 'vuelidate';
import * as actionTypes from '../../src/store/actions.type';

jest.mock('../../src/store/actions');

const localVue = createLocalVue();

localVue.use(VueTailwind);
localVue.use(Vuex);
localVue.use(Vuelidate);

let wrapper;
const propsData = {
  category: {
    name: 'Test category',
    favorite_things: []
  },
  singularizedCategoryName: 'Test category'
}

describe('NewFavoriteThing component', () => {
  beforeEach(() => {
    wrapper = shallowMount(NewFavoriteThing, {
      localVue,
      store: new Vuex.Store({
        actions
      }),
      propsData,
    });
  })
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleAddNewFavoriteThing method', () => {
    it('should validate form and touch all fields if form is invalid', async () => {
      wrapper.vm[actionTypes.ADD_FAVORITE_THING] = jest.fn();
      await wrapper.vm.handleAddNewFavoriteThing()

      expect(wrapper.vm.submitting).toBe(false);
      expect(wrapper.vm[actionTypes.ADD_FAVORITE_THING]).not.toBeCalled()
    })

    it('should call add favorite thing action if form is valid', async () => {
      const newFavoriteThing = {
        id: 1,
        title: 'test',
        ranking: 1,
        category: 1,
        metadata: []
      }
      wrapper.vm.newFavoriteThing = newFavoriteThing
      wrapper.vm[actionTypes.ADD_FAVORITE_THING] = jest.fn();
      await wrapper.vm.handleAddNewFavoriteThing();

      expect(wrapper.vm[actionTypes.ADD_FAVORITE_THING]).toBeCalledWith(newFavoriteThing);
    })
  })

  describe('handleEditNewFavoriteThing method', () => {
    it('should validate form and touch all fields if form is invalid', async () => {
      wrapper.vm[actionTypes.UPDATE_FAVORITE_THING] = jest.fn();
      await wrapper.vm.handleEditNewFavoriteThing()

      expect(wrapper.vm.submitting).toBe(false);
      expect(wrapper.vm[actionTypes.UPDATE_FAVORITE_THING]).not.toBeCalled()
    })

    it('should call add favorite thing action if form is valid', async () => {
      const newFavoriteThing = {
        id: 1,
        title: 'test',
        ranking: 1,
        category: 1,
        metadata: []
      }
      wrapper.vm.newFavoriteThing = newFavoriteThing
      wrapper.vm[actionTypes.UPDATE_FAVORITE_THING] = jest.fn();
      await wrapper.vm.handleEditNewFavoriteThing();

      expect(wrapper.vm[actionTypes.UPDATE_FAVORITE_THING]).toBeCalledWith({
        favoriteThingId: newFavoriteThing.id,
        updatedFavoriteThingData: newFavoriteThing
      });
    });
  });

  describe('handleMetadataChange', () => {
    it('should update new favorite thing metadta', () => {
      const updatedMetadata = [{ name: 'update' }];
      wrapper.vm.handleMetadataChange(updatedMetadata);

      expect(wrapper.vm.newFavoriteThing.metadata).toEqual(updatedMetadata);
    })
  });

  describe('transformMetadata', () => {
    it('should convert vue tag input to flat array if metadata is an array', () => {
      wrapper.vm.newFavoriteThing.metadata = [
        {
          type: 'list',
          label: 'sports',
          value: ['Basketball']
        }
      ]
      expect(wrapper.vm.transformMetadata()).toEqual([
        {
          type: 'list',
          label: 'sports',
          value: [{
            key: '',
            value: 'Basketball'
          }]
        }
      ])
    })

    it('should convert vue tag input to flat array if metadata is a string', () => {
      wrapper.vm.newFavoriteThing.metadata = JSON.stringify([
        {
          type: 'list',
          label: 'sports',
          value: ['Basketball']
        }
      ]);

      expect(wrapper.vm.transformMetadata()).toEqual([
        {
          type: 'list',
          label: 'sports',
          value: [{
            key: '',
            value: 'Basketball'
          }]
        }
      ])
    })
  })

  describe('rankingInfo computed field', () => {
    it('should return empty string if rank is currentrank', () => {
      wrapper.vm.newFavoriteThing.ranking = 3
      wrapper.setProps({
        favoriteThing: {
          ranking: wrapper.vm.newFavoriteThing.ranking
        }
      });

      expect(wrapper.vm.rankingInfo).toEqual('');
    });
    it('should return rank above if current ranking is in the middle', () => {
      wrapper.setProps({
        category: {
          ...propsData.category,
          favorite_things: [
            {
              title: 'test',
              ranking: 1
            },
            {
              title: 'second test',
              ranking: 2
            },
            {
              title: 'third test',
              ranking: 3
            }
          ]
        }
      });

      wrapper.vm.newFavoriteThing.ranking = 2;

      expect(wrapper.vm.rankingInfo).toEqual('Rank above second test')
    })
    it('should return rank as least if current ranking is the least', () => {
      wrapper.setProps({
        category: {
          ...propsData.category,
          favorite_things: [
            {
              title: 'test',
              ranking: 1
            },
            {
              title: 'second test',
              ranking: 2
            },
            {
              title: 'third test',
              ranking: 3
            }
          ]
        }
      });

      wrapper.vm.newFavoriteThing.ranking = 4;

      expect(wrapper.vm.rankingInfo).toEqual('Rank as least favorite Test category');
    })
  });

  it('should set newFavoriteThing to favoriteThing if passed as props on mount', () => {
    wrapper = shallowMount(NewFavoriteThing, {
      localVue,
      store: new Vuex.Store({
        actions
      }),
      propsData: {
        ...propsData,
        favoriteThing: {
          title: 'test',
          ranking: 1,
          metadata: []
        },
      },
    });

    expect(wrapper.vm.newFavoriteThing).toEqual(wrapper.vm.favoriteThing)
  })
});
