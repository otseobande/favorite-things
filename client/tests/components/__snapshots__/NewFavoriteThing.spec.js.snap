// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFavoriteThing component should match snapshot 1`] = `
<div class="text-gray-700">
  <h2 class="text-lg font-bold"><span>
        Add new favorite <span class="lowercase">Test category</span></span></h2>
  <form class="mt-5">
    <div><label class="font-bold">Title</label>
      <t-input-stub name="title" value="" placeholder="Enter favorite Test category title" type="text" baseclass="border block rounded" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" class="h-10 w-full"></t-input-stub>
      <!---->
    </div>
    <div class="mt-2"><label class="font-bold">Description </label> <span class="text-sm italic text-gray-500">(optional)</span>
      <t-textarea-stub name="description" value="" placeholder="Describe favorite Test category" rows="2" wrap="soft" baseclass="border block rounded" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" class="w-full"></t-textarea-stub>
      <!---->
    </div>
    <div class="mt-2"><label class="font-bold">Ranking</label>
      <div>
        <t-input-stub name="ranking" value="1" max="1" min="1" type="number" baseclass="border block rounded" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" class="h-10 mr-0 inline"></t-input-stub> <small class="ml-2 capitalize">Rank as favorite Test category</small>
      </div>
      <!---->
    </div>
    <additionalinformation-stub currentmetadata="" editting="true"></additionalinformation-stub>
    <div class="mt-8 flex">
      <t-button-stub tagname="button" type="submit" variant="primary" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm" class="mr-2 font-bold"><span>
            Save favorite<span class="lowercase">&nbsp;Test category</span></span></t-button-stub>
      <t-button-stub tagname="button" type="button" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm" class="font-bold">
        Cancel
      </t-button-stub>
    </div>
  </form>
</div>
`;
